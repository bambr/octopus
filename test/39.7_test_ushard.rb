#!/usr/bin/ruby
# coding: utf-8

$: << File.dirname($0)
require '39_test_ushard'

$one_env.meta 'shard 1 alter por one'
$one_env.meta 'shard 1 obj_space 0 create tree unique string 0'
$one.insert [1,"one"], :shard => 1
$two_env.meta 'shard 1 alter por one two'
sleep 0.1

# switch master
$two_env.meta 'shard 1 alter por two one'
sleep 0.1


$one.insert [1,"one2"], :shard => 1
$one.select 1, :shard => 1
$two.select 1, :shard => 1
$two.insert [1,"two2"], :shard => 1
sleep 0.1
$one.select 1, :shard => 1
$two.select 1, :shard => 1

$two_env.env_eval do
  stop
  puts `./octopus --cat 00000000000000000001.snap`.gsub(/ tm:\d+(\.\d+)? /, ' ')
  puts `./octopus --cat 00000000000000000002.xlog`.gsub(/ tm:\d+(\.\d+)? /, ' ')
  puts `grep META octopus.log`.gsub(/^\d+\.\d+ \d+ /, '')
end

