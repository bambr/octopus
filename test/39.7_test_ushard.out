# one.ping()
:pong

# two.ping()
:pong

# one.meta(shard 1 create por)
ok

# one.meta(shard 1 obj_space 0 create tree unique string 0)
ok

# one.insert([1, "one"], {:shard=>1})
1

# two.meta(shard 1 add_replica two)
ok

# two.meta(shard 1 master two)
ok

# one.insert([1, "one2"], {:shard=>1})
2

# one.select(1, {:shard=>1})
[["\x01\x00\x00\x00", "one2"]]

# two.select(1, {:shard=>1})
[["\x01\x00\x00\x00", "one2"]]

# two.insert([1, "two2"], {:shard=>1})
2

# one.select(1, {:shard=>1})
[["\x01\x00\x00\x00", "two2"]]

# two.select(1, {:shard=>1})
[["\x01\x00\x00\x00", "two2"]]

lsn:2 scn:-1 t:sys/snap_initial ver:0 count:1 flags:0x00000000
lsn:2 shard:1 scn:4 t:sys/shard_create SHARD_CREATE shard_id:1 POR Box count:1 run_crc:0xc50dacae master:one repl:two
lsn:2 shard:1 scn:4 t:snap/usr240 CREATE_OBJECT_SPACE n:0 flags:00000003 cardinalty:0 PK: i:0 min_tuple_cardinality:0 cardinality:1 type:5 unique:1 field0:{index:0 type:7 sort:1}
lsn:2 shard:1 scn:4 t:snap/snap_data n:0 <1:"\x01\x00\x00\x00", "one">
lsn:2 shard:1 scn:4 t:sys/shard_final 
lsn:2 scn:-1 t:sys/snap_final 
lsn:2 shard:1 scn:4 t:sys/shard_alter SHARD_ALTER shard_id:1 POR Box count:1 run_crc:0xb4c41ab4 master:one repl:two
lsn:3 shard:1 scn:5 t:sys/shard_alter SHARD_ALTER shard_id:1 POR Box count:1 run_crc:0xb4c41ab4 master:two repl:one
lsn:4 shard:1 scn:6 t:wal/usr13 INSERT n:0 flags:00000000 <1:"\x01\x00\x00\x00", 845508207:"one2">
lsn:5 shard:1 scn:7 t:wal/usr13 INSERT n:0 flags:00000000 <1:"\x01\x00\x00\x00", 846165876:"two2">
238/load shard {1} I> META mode: LOADING => shard_add
238/load shard {1} I> META mode: PARTIAL_PROXY, SCN: 4, type: POR, peer: [one, two], proxy_addr: 'one/127.0.0.1:33013' => hot_standby/one/127.0.0.1:33013/init
239/remote_hot_standby {1} I> META mode: PARTIAL_PROXY, SCN: 4, type: POR, peer: [one, two], proxy_addr: 'one/127.0.0.1:33013' => hot_standby/127.0.0.1:33013/connect
238/load shard {1} I> META mode: PARTIAL_PROXY, SCN: 4, type: POR, peer: [one, two], proxy_addr: 'one/127.0.0.1:33013' => wal_final_row
238/load shard {1} I> META mode: PARTIAL_PROXY, SCN: 4, type: POR, peer: [one, two], proxy_addr: 'one/127.0.0.1:33013' => wal_final_row
239/remote_hot_standby {1} I> META mode: PARTIAL_PROXY, SCN: 4, type: POR, peer: [one, two], proxy_addr: 'one/127.0.0.1:33013' => hot_standby/127.0.0.1:33013/ok
239/remote_hot_standby {1} I> META mode: PARTIAL_PROXY, SCN: 4, type: POR, peer: [one, two], proxy_addr: 'one/127.0.0.1:33013' => wal_final_row
239/remote_hot_standby {1} I> META mode: LOCAL, SCN: 4, type: POR, peer: [two, one] => route update, force notify
239/remote_hot_standby {1} I> META mode: LOCAL, SCN: 4, type: POR, peer: [two, one] => primary
